test/test_functionalization.py

```diff
@@ -0,0 +1,215 @@
+import torch
+from torch.testing._internal.common_utils import TestCase, run_tests
+from torch.testing._internal.logging_tensor import LoggingTensor, capture_logs, log_input
+
+def maybe_log_functional_input(name, handle, is_logging):
+    if is_logging:
+        assert torch._is_functional_tensor(handle)
+        log_input(name, torch._from_functional_tensor(handle))
+
+def fooo():
+    print("ENABLE")
+    torch._enable_functionalization()
+
+def barr():
+    print("DISABLE")
+    torch._disable_functionalization()
+
+
+class TestFunctionalization(TestCase):
+
+    def assert_functionalization(self, func, inpt):
+        input_clone = inpt.clone()
+
+        input_clone2 = inpt.clone()
+        input_clone_logging = LoggingTensor(inpt.clone())
+
+        input_functional = torch._to_functional_tensor(input_clone2)
+        input_functional_logging = torch._to_functional_tensor(input_clone_logging)
+
+        # Runs expecttests for LoggingTensor output.
+        torch._enable_functionalization()
+        func(input_functional_logging, logging=True)
+        torch._disable_functionalization()
+
+        # Compare outputs (and mutated inputs), with and without functionalization.
+        out_ref = func(inpt, logging=False)
+
+        torch._enable_functionalization()
+        out_functional = func(input_functional, logging=False)
+        torch._disable_functionalization()
+
+        # We need to sync the input tensors first, in case there are any queued mutations left.
+        torch._sync(input_functional)
+        torch._sync(out_functional)
+        self.assertEqual(out_ref, torch._from_functional_tensor(out_functional))
+        self.assertEqual(inpt, torch._from_functional_tensor(input_functional))  # input mutations should still occur
+
+    def test_simple(self):
+        def f(x, *, logging: bool):
+            # simple test: 1 view op, 1 inplace op
+            with capture_logs() as logs:
+                maybe_log_functional_input("x", x, logging)
+                tmp = torch.ones(4, 2)
+                y = x.view(4, 2)
+                y.add_(tmp)
+                z = x * x
+            if logging:
+                self.assertExpectedInline('\n'.join(logs), """\
+$0 = input('x')
+$1 = torch._ops.aten.view($0, [4, 2])
+$2 = torch._ops.aten.add($1, tensor([[1., 1.],
+        [1., 1.],
+        [1., 1.],
+        [1., 1.]]))
+$3 = torch._ops.aten.view($2, [4, 2])
+$4 = torch._ops.aten.mul($3, $3)""")
+            else:
+                return y
+        x = torch.ones(4, 2)
+        self.assert_functionalization(f, x)
```

Comment:
Totally agreed. I originally wrote it that way to make sure that the expecttest output was next to the actual function for ease of comparison, but your way looks cleaner and accomplishes the same thing.

Reply:
oh, I tried it out and I don't actually think that this works. See https://github.com/ezyang/expecttest/blob/4be7413c98b4c6bdb4b26bfc8358c0a909c8d675/expecttest/__init__.py#L197 - the expected string argument to `assertExpectedInline()` needs to be a string literal that the expecttests can update in place- otherwise it won't know where the `expected_logs` variable came from.

Reply:
I agree with your assessment, that is unfortunate. How about the following instead? We could have `assert_functionalization` return the logs:
```py
def f(x):
  tmp = torch.ones(4, 2)
  y = x.view(4, 2)
  y.add_(tmp)
  z = x * x
  return y

x = torch.ones(4, 2)
logs = self.run_functionalization_checks(f, x, return_logs=True)
self.assertExpectedInline('\n'.join(logs), """
...
""")
```

Reply:
(if that doesn't work out -- let's just ship this, we can always figure out how to make it better later...)